test/e2e/annotations/canary.go

```diff
@@ -86,6 +86,54 @@ var _ = framework.DescribeAnnotation("canary-*", func() {
 				NotContains(canaryService)
 		})
 
+		// issue: https://github.com/kubernetes/ingress-nginx/issues/9944
+		// canary routing should not overwrite custom errors
+		ginkgo.It("should respond with a 401 status from the custom errors backend when canary responds with a 401",
+			func() {
+				host := "foo"
+				annotations := map[string]string{
+					"nginx.ingress.kubernetes.io/custom-http-errors": "401",
+					"nginx.ingress.kubernetes.io/default-backend":    framework.DefaultBackendService,
+				}
+
+				f.NewDefaultBackendDeployment()
+
+				f.EnsureIngress(framework.NewSingleIngress(
+					host,
+					"/",
+					host,
+					f.Namespace,
+					framework.HTTPBunService,
+					80,
+					annotations,
+				))
+
+				f.WaitForNginxServer(host, func(server string) bool {
+					return strings.Contains(server, "server_name foo")
+				})
+
+				f.EnsureIngress(framework.NewSingleIngress(
+					canaryService,
+					"/",
+					host,
+					f.Namespace,
+					canaryService,
+					80,
+					map[string]string{
+						"nginx.ingress.kubernetes.io/canary":           "true",
+						"nginx.ingress.kubernetes.io/canary-by-header": "CanaryByHeader",
+					},
+				))
+
+				f.HTTPTestClient().
+					GET("/status/401").
+					WithHeader("Host", host).
+					Expect().
+					Status(http.StatusUnauthorized).
+					Body().
+					Contains("401")
```

Comment:
Due to the passage of time, I need to reload some memories.

Can this case cover the actual scenarios in that issue?

Reply:
Yes, the test case was written specifically to reproduce the issue, before fixing it

Reply:
Means we are good to go? ;)

Reply:
I have been trying to find time to verify what would happen with this test case before the Lua part modification.


This is the only issue that I am currently worried about.

Reply:
@dfl-aeb @Spazzy757 Sorry for the delay, very busy with daily work.

Can you help me verify the result of this test case when there are no modifications to the Lua part? Other than that, I have no other concerns about this PR.

Reply:
@Spazzy757 can you help out here? :)

Reply:
Sorry to bother you again @Spazzy757 
Do you think you can find some time in the next few weeks?

Reply:
I can try do this next week and paste the results, I wrote the test before making changes to the Lua to verify the issue.

I also tested this, manually using the same method I used in https://github.com/kubernetes/ingress-nginx/issues/9944

Reply:
Do you have any updates @Spazzy757 ?

Reply:
@Spazzy757 @tao12345666333 friendly reminder :)

Reply:
@Spazzy757 could you please share your test results?